### PR TITLE
Make colormap handling consistent and allow discrete mapping

### PR DESCRIPTION
**Pull request by philippjfr**
This PR makes colormapping more consistent between backends:

* Adds support for ``color_levels`` to generate discrete colormaps ( https://github.com/ioam/holoviews/issues/1856)
* Allows defining an explicit mapping from category to color.
* Adds the symmetric plot option to bokeh plots for consistency with matplotlib backend
* Removes hard dependency on matplotlib in bokeh backend by defining ``MPL_COLORS`` in holoviews (https://github.com/ioam/holoviews/issues/829, https://github.com/ioam/holoviews/issues/1807)
* Adds support for colormap interpolation to bokeh backend so, even in the absence of matplotlib, colormaps are smoothly interpolated and therefore (almost) identical to matplotlib (https://github.com/ioam/holoviews/issues/1502) (there are some rounding issues causing, at least for me, imperceptible differences)
* Adds a new ``Styling_Plots`` user guide which currently covers Cycles/Palettes and colormaps.
* Enables 'transparent' as an alias for transparent NaN color and sets Raster/Image/QuadMesh default NaN color to transparent (https://github.com/ioam/holoviews/issues/2487)
* Extended the default color Cycle (https://github.com/ioam/holoviews/issues/1591)

This means that the following list of colormaps are now supported consistently across backends:

```
['Blues',
 'BrBG',
 'BuGn',
 'BuPu',
 'GnBu',
 'Greens',
 'Greys',
 'OrRd',
 'Oranges',
 'PRGn',
 'PiYG',
 'PuBu',
 'PuBuGn',
 'PuOr',
 'PuRd',
 'Purples',
 'RdBu',
 'RdGy',
 'RdPu',
 'RdYlBu',
 'RdYlGn',
 'Reds',
 'Spectral',
 'YlGn',
 'YlGnBu',
 'YlOrBr',
 'YlOrRd',
 'Accent',
 'Dark2',
 'Paired',
 'Pastel1',
 'Pastel2',
 'Set1',
 'Set2',
 'Set3',
 'magma',
 'inferno',
 'plasma',
 'viridis']
```

I'm not opposed to adding colorcet to that list, but it's now no longer required for our colormap handling to be consistent.

- [x] Add unit tests

### CODE REVIEW

**Comment by philippjfr** [comment]
@jbednar @jlstevens Would be good to get your thoughts on this.

**Comment by jbednar** [comment]
I'd favor adding colorcet, but I'm not sure my vote should be the deciding factor.

**Comment by philippjfr** [comment]
>I'd favor adding colorcet

That is pretty much entirely orthogonal to this PR although I'm happy to add them. At least now they should appear identical in matplotlib and bokeh. What fraction of colorcet colormaps now have readable names btw? You can already use colorcet colormaps with HoloViews by passing them in explicitly all "adding it" means that you can reference them by name, which isn't very useful with obscure names.

**Comment by jbednar** [comment]
Out of the 51 distinct colormaps in colorcet, 21 have human-friendly names:

```
bgy
bgyw
bjy
bkr
bky
blues
bmw
bmy
colorwheel
coolwarm
dimgray
fire
gray
gwv
isolum
kb
kbc
kg
kgy
kr
rainbow
```

Or if you count the `_r` variants of each of each one, there are 102 maps and 42 with human-friendly names.

**Comment by jbednar** [comment]
It's nice to be able to pass them in by name like this, but it's also nice to be sure that you are using perceptually uniform maps, and that's hard to tell from the names (plus inferno, plasma, viridis, and a few other common uniform ones. I guess we could provide a flag people could set "warn_for_perceptually_nonuniform_colormaps"?

**Comment by philippjfr** [comment]
>Out of the 51 distinct colormaps in colorcet, 21 have human-friendly names:

Okay, that definitely seem sufficient to include, at least when colorcet is available. No opinion on requiring it.

>It's nice to be able to pass them in by name like this, but it's also nice to be sure that you are using perceptually uniform maps, and that's hard to tell from the names.

That's true, I think documenting them well is the most important thing. I'm not opposed to adding an option to warn but I'm not sure how discoverable it would be in practice.

**Comment by philippjfr** [comment]
Here is the full list including named colorcet maps, some are duplicated because they have different names in bokeh and matplotlib, but I've tried to smooth over those differences as much as possible:

screen shot 2018-03-27 at 1 32 51 pm

**Comment by philippjfr** [comment]
Perhaps more helpful if split up into three sections:

## Matplotlib 

screen shot 2018-03-27 at 1 49 59 pm

## Bokeh

screen shot 2018-03-27 at 1 50 35 pm

## Colorcet

screen shot 2018-03-27 at 1 50 42 pm

File: holoviews/plotting/mpl/element.py [hunk 1]
```diff
@@ -658,6 +652,14 @@ def _norm_kwargs(self, element, ranges, opts, vdim, prefix=''):
                     alpha = int(color[-2:], 16)/255.
                     color = color[:-2]
                 colors[k] = {'color': color, 'alpha': alpha}
+
+        if isinstance(cmap, dict):
+            factors = np.unique(values)
+            palette = [cmap.get(f, colors.get('NaN', {'color': '#8b8b8b'})['color'])
```
Comment (jlstevens) [inline holoviews/plotting/mpl/element.py]: What is this magic value '#8b8b8b'? I'm guessing it is some grayish color...

Comment (philippjfr) [reply holoviews/plotting/mpl/element.py]: Right, can probably specify that somewhere else.

Comment (jbednar) [reply holoviews/plotting/mpl/element.py]: Is that a `default_nan_color` or `color_for_missing_values`?

**Comment by jbednar** [comment]
> Perhaps more helpful if split up into three sections:

Splitting it up is definitely important, and it *could* be useful to split it by backend, but I would think that a categorical distinction would be much more useful (linear (aka sequential), grey, diverging, cyclic, qualitative, isoluminant) the way matplotlib and colorcet do.  If you give me your code for rendering the colorbars above, I'd be happy to categorize all the colormaps in this way, using https://matplotlib.org/tutorials/colors/colormaps.html and colorcet's naming scheme to establish the categories.  And then if there were a way to mark which ones are perceptually uniform, I think this could be a very good reference for helping people choose colormaps.

**Comment by philippjfr** [comment]
>Splitting it up is definitely important, and it could be useful to split it by backend

This split was for illustration purposes on this PR, not for documentation. I'd be happy to list them separately. If you're willing to take that on I'd suggest a second argument to the new ``list_colormaps``, which filters the colormaps by type. You can then update the Styling Plots user guide added in this PR to list each type separately.

File: holoviews/plotting/mpl/element.py (hunk 1 above)
Comment (philippjfr) [reply holoviews/plotting/mpl/element.py]: It's really ``color_for_values_which_have_not_been_assigned_a_explicit_color``.

File: holoviews/plotting/bokeh/element.py [hunk 2]
```diff
@@ -1046,7 +1046,8 @@ def _get_colormapper(self, dim, element, ranges, style, factors=None, colors=Non
             low, high = None, None
 
         cmap = colors or style.pop('cmap', 'viridis')
-        nan_colors = {k: rgba_tuple(v) for k, v in self.clipping_colors.items()}
+        nan_colors = {k: (0, 0, 0, 0) if v == 'transparent' else rgba_tuple(v)
+                      for k, v in self.clipping_colors.items()}
```
[the remark this answers is not in the document]
Comment (jlstevens) [reply holoviews/plotting/bokeh/element.py]: There ought to be an HTML color than means this imho!

Sadly, I don't think there is one though...

[the remark this answers is not in the document]
Comment (jbednar) [reply holoviews/plotting/bokeh/element.py]: Adding the check to rgba_tuple won't work?

[the remark this answers is not in the document]
Comment (philippjfr) [reply holoviews/plotting/bokeh/element.py]: Sure, can do, I could also add it to the ``MPL_COLORS`` dictionary and perhaps name it something more generic like ``COLOR_ALIASES``. Currently it just defines the matplotlib color shorthands b, c, g, k, m, r, w, y.

**Comment by philippjfr** [comment]
@jbednar Would you like to do the colormap categorization in this PR or do you want to follow up in another later?

**Comment by philippjfr** [comment]
From my perspective this is now ready to review, but @jbednar will have to say if he wants to list the colormaps in groups in this PR.

**Comment by philippjfr** [comment]
Although I suppose we could also extend the color cycle as suggested in https://github.com/ioam/holoviews/issues/1591 in this PR. The final choice was:

```
colors = ['#30a2da', '#fc4f30', '#e5ae38', '#6d904f', '#8b8b8b', '#17becf',
          '#9467bd', '#d62728', '#1f77b4', '#e377c2', '#8c564b', '#bcbd22']
```

screen shot 2018-03-28 at 1 52 55 am

**Comment by jbednar** [comment]
I can make another pr; this one can be merged.

**Comment by philippjfr** [comment]
Okay, had to resolve some final issues with mpl and bokeh colormap consistency but I've not added unit tests so this is definitely ready for review.

**Comment by jbednar** [comment]
I'm editing the Styling Plots notebook now, so please do review but probably best to hold off from merging until I'm done.

**Comment by philippjfr** [comment]
Edits look good to me, over to @jlstevens for review.

**Comment by jbednar** [comment]
I'm done making changes except that I'll still try to pull out the colormaps into different categories in the styling notebook.  Merging doesn't have to wait on that, though; it should be a minor and very local change.

**Comment by philippjfr** [comment]
>The changes all look good to me in the latest version, apart from wondering why clipping_colors is defined in so many separate places.

Mainly because it's appropriate for raster-like types but not anything else, e.g. mapping points, vectors to be transparent is obviously not correct, so we can't just declare it on the baseclass.

**Comment by jbednar** [comment]
Ok. Looks like fixing the four different places it's currently redefined would require a new base class RasterlikePlot under ColorbarPlot, and I guess it's not worth that.

**Comment by philippjfr** [comment]
Right, and we'd either have to define a mixin class or separate subclasses for each backend.

**Comment by jlstevens** [comment]
Looks good to me. Merging.

**Pull request merged**